lib/Core/SpecialFunctionHandler.cpp

```diff
@@ -936,48 +1010,133 @@ void SpecialFunctionHandler::handleMarkPersistent(ExecutionState &state,
 
   if (name.length() == 0) {
     name = "unnamed";
-    klee_warning("klee_pmem_mark_persistent: renamed empty name to \"unnamed\"");
+    klee_warning("klee_pmem_mark_persistent: name parameter empty, using \"unnamed\"");
+  }
+
+  ref<Expr> addr = arguments[0];
+  ref<Expr> size = arguments[1];
+  uint64_t realSize = 0;
+  if (ConstantExpr *ce = dyn_cast<ConstantExpr>(size.get())) {
+    realSize = ce->getZExtValue();
+  } else {
+    size->dump();
+    klee_error("Not sure how to handle symbolic size argument yet!");
   }
 
   Executor::ExactResolutionList rl;
-  executor.resolveExact(state, arguments[0], rl, "mark_persistent");
+  for (uint64_t offset = 0; offset < realSize; offset += PersistentState::MaxSize) {
+    ref<Expr> offsetExpr = ConstantExpr::create(offset, Expr::Int64);
+    ref<Expr> ptrExpr = AddExpr::create(addr, offsetExpr);
+    executor.resolveExact(state, ptrExpr, rl, "mark_persistent");
+  }
   
   for (Executor::ExactResolutionList::iterator it = rl.begin(), 
          ie = rl.end(); it != ie; ++it) {
     const MemoryObject *mo = it->first.first;
     mo->setName(name);
+    // klee_warning("Marking object at %p as persistent", (void*)mo->address);
     
     ExecutionState *s = it->second;
-
     // FIXME: Type coercion should be done consistently somewhere.
     bool res;
+    // bool success __attribute__ ((unused)) =
+    //   executor.solver->mustBeTrue(*s, 
+    //                               EqExpr::create(ZExtExpr::create(arguments[1],
+    //                                                               Context::get().getPointerWidth()),
+    //                                              mo->getSizeExpr()),
+    //                               res);
     bool success __attribute__ ((unused)) =
       executor.solver->mustBeTrue(*s, 
-                                  EqExpr::create(ZExtExpr::create(arguments[1],
+                                  EqExpr::create(ConstantExpr::create(PersistentState::MaxSize,
                                                                   Context::get().getPointerWidth()),
                                                  mo->getSizeExpr()),
                                   res);
     assert(success && "FIXME: Unhandled solver failure");
     
     if (res) {
       executor.executeMarkPersistent(*s, mo);
+      // Just return the same address as what we received.
+      // -- This may get re-executed, but technically "s" can be a different state.
+      executor.bindLocal(target, *s, addr);
     } else {      
       executor.terminateStateOnError(*s, 
                                      "wrong size given to klee_pmem_mark_persistent[_name]", 
                                      Executor::User);
     }
   }
+
+}
+
+void SpecialFunctionHandler::handleIsPmem(ExecutionState &state,
```

Comment:
What do you use this for?

Reply:
oh, probably to check that mmap calls result in pmem?

Reply:
Yup. I use it in parts of the POSIX runtime emulation.